Add SimpleMerkleTree

ernestognw opened the pull request:
### Description

Alternative to #31 with less repeated code.
The SimpleMerkleTree also receives `BytesLikes` leaves so that the input is consistent with StandardMerkleTree and we can use inheritance to avoid code repetition.

I removed the `T extends BytesLike[]` for simplicity and because `BytesLike[]` is the only input supported.

Note that StandardMerkleTree overloaded functions are for backwards compatibility. We can get rid of them if we release a new non-backwards compatible version

Amxx commented on src/bytes.ts:
```diff
@@ -1,19 +1,20 @@
+import type { BytesLike, Hexable } from "@ethersproject/bytes";
 
-export type Bytes = Uint8Array;
+type Hex = BytesLike | Hexable | number | bigint;
```
I think this should have a different name. Its `Hexlifyable`, or `Hexable`, but not `Hex`

Amxx replied:
Note that, while its true that `toHex` supports more types than just `BytesLike`, it feels to me like this wyde `Hex` type is not used anywhere but here.

I would restrict `compare` to `BytesLike` because that is all that we actually need in this repo, and remove that type definition.

Amxx commented on src/bytes.ts:
```diff
@@ -1,19 +1,20 @@
+  arrayify as toBytes,
+  hexlify as toHex,
+  concat,
+  isBytes,
```
I don't think this is never used. It feels like `isBytesLike` is enough.

ernestognw replied:
Right, I'll remove it

Amxx commented on src/standard.ts:
```diff
@@ -1,202 +1,196 @@
-    for (const [leafIndex, { valueIndex }] of hashedValues.entries()) {
-      indexedValues[valueIndex]!.treeIndex = tree.length - leafIndex - 1;
-    }
+export function standardLeafHash<T extends BytesLike[] = BytesLike[]>(
```
Is `T extends BytesLike[]` ok? The values could be numbers, arrays, structures, basically anything supported by the AbiEncoder.

Amxx commented on src/standard.ts:
```diff
@@ -1,202 +1,196 @@
+export type StandardMerkleTreeValue = SimpleMerkleTreeValue;
 
-    const tree = makeMerkleTree(hashedValues.map(v => v.hash));
+export type StandardMerkleTreeData = SimpleMerkleTreeData & {
```
`SimpleMerkleTreeData` is 
```typescript
export type SimpleMerkleTreeValue = {
  value: BytesLike[];
  treeIndex: number;
};
export type SimpleMerkleTreeData = {
  format: string;
  tree: HexString[];
  values: SimpleMerkleTreeValue[];
};
```
which means that the proposed `StandardMerkleTreeData` uses the values to be `HexString[]`.

Note that the current type is:
```typescript
interface StandardMerkleTreeData<T extends any[]> {
  format: 'standard-v1';
  tree: string[];
  values: {
    value: T;
    treeIndex: number;
  }[];
  leafEncoding: string[];
}
``` 
Where the values are `any[]`.

This means that the proposed change is breaking backward compatibility, particularly is loses the ability to load old dump. That is a big no to me.

ernestognw replied:
> This means that the proposed change is breaking backward compatibility, particularly is loses the ability to load old dump. 

I was completely unfamiliar with this. It's also a no for me. I'll mark it as draft before reviewing your iteration and updating this one with `any` type.

Amxx commented:
There might be less code duplication, but I also think there is more code complexity. I'm not sure we are gaining much.

Also, we would need to find a way to support any[] values. Elements of the leafs are not all `BytesLike`.
For example, we must be able to do 
```typescript
    const leaves = [
      [0, []],
      [1, ['openzeppelin']],
      [2, ['hello', 'world']],
      [3, ['merkle', 'tree']],
    ];
    const types = [ 'uint256', 'string[]' ];
    const tree = StandardMerkleTree.of(leaves, types);
```

Amxx commented:
There is also something that is bothering me with the way options are stored in the trees.

If possible I'd like to avoid that. I get that its necessary to set the options in `SimpleMerkleTree` so that hashleaf in the `StandardMerkleTree` works as part of the `SimpleMerkleTree`'s construction (so before we can set a leafEncoding value in the standard tree ... but maybe there is another way

Amxx commented on src/simple.ts:
```diff
@@ -0,0 +1,255 @@
+    protected readonly values: SimpleMerkleTreeValue[],
+    protected readonly options: SimpleMerkleTreeOptions = {}
+  ) {
+    this.options = Object.assign(options, defaultOptions);
```
We don't want to override options that are set with the default ones.
```suggestion
    this.options = Object.assign(defaultOptions, options);
```

frangio replied:
@Amxx's comment is right but there is another (small) issue with `Object.assign`. TypeScript allows the user to pass in `{ sortLeaves: undefined }` and this will override the default. AFAIK the best way to solve this is to write out the full object with each option listed:

```suggestion
    this.options = {
      sortLeaves: options.sortLeaves ?? defaultOptions.sortLeaves,
      /* ... the rest */
    }
```

Amxx commented:
Another think I just realized is broken in this PR is the interface of `getMultiProof(leaves: (number | T)[]): MultiProof<string, T>`

Amxx commented:
I iterated on this PR in https://github.com/OpenZeppelin/merkle-tree/tree/support-raw-leaves-inheritance-amxx

Main differences:
- Interface `MerkleTree<T>` and class `MerkleTree<T>` provide a generic base for all tree types (`T` is the leaf type)
- `SimpleMerkleTree extends MerkleTree<BytesLike>`
- `StandardMerkleTree<T extends any[]> extends MerkleTree<T>`
- Re-introduce support for non BytesLike element in the leaf (with corresponding tests)
- Reduce the number of variants to reduce code (this includes removing putting leafEncoding in the options)
- Re-add the generic multiproof to keep support for `Multiproof<T>` in `StandardMerkleTree`

I think this is looking better than #31

ernestognw commented on src/simple.ts:
```diff
@@ -0,0 +1,32 @@
+}
+
+export class SimpleMerkleTree extends MerkleTreeImpl<BytesLike> {
+  static of = (values: BytesLike[], options: MerkleTreeOptions = {}): SimpleMerkleTree => {
+    const [tree, indexedValues] = MerkleTreeImpl.prepare(values, options, formatLeaf);
+    return new SimpleMerkleTree(tree, indexedValues, formatLeaf);
+  };
```
I guess there's no need to keep it as a lambda:

```suggestion
  static of(values: BytesLike[], options: MerkleTreeOptions = {}): SimpleMerkleTree {
    const [tree, indexedValues] = MerkleTreeImpl.prepare(values, options, formatLeaf);
    return new SimpleMerkleTree(tree, indexedValues, formatLeaf);
  }
```

ernestognw commented:
I reviewed the updates and generally addressed comments because I think they're not too controversial. Added explanations accordingly.

The main differences are:

1. Each `StandardMerkleTree` and `SimpleMerkleTree` now define their `format` on their own data types. Also `MerkleTree` was marked as abstract so that `dump()` remains virtual until implemented.
2. Changed the leafHasher function in the StandardMerkleTree to an arrow function. I think the alternative `bind` is confusing because it specifies `this` as null.
3. Specialized the errors into `InvalidArgumentError` and `InvariantError`. I noticed there was a lot of `new Error` around when we already had a `throwError` helper.

At this point, the PR looks good to me. I left a couple of open comments that need comments. Also, appreciate feedback on point number 3 regarding errors. If that's too much to discuss here let's revert to [ada28f1](https://github.com/OpenZeppelin/merkle-tree/pull/36/commits/ada28f1bb405ca688f1a9787a58cf8e5fb870669)

Amxx commented on src/merkletree.ts:
```diff
@@ -0,0 +1,191 @@
+    });
+  }
+
+  protected validateValue(valueIndex: number): HexString {
```
That works for me.

Amxx commented on src/standard.ts:
```diff
@@ -1,213 +1,72 @@
-    treeIndex: number;
-  }[];
+export type StandardMerkleTreeData<T extends any[]> = MerkleTreeData<T> & {
   leafEncoding: string[];
```
I like having `format: string` in `MerkleTreeData<T>` and override that with the more restrictive `fomat: `stnadard-v1` in `StandardMerkleTreeData<T extends any[]>`
👍🏻

Amxx commented on src/standard.ts:
```diff
@@ -1,213 +1,71 @@
-      throw new Error(`Unknown format '${data.format}'`);
-    }
-    return new StandardMerkleTree(data.tree.map(hexToBytes), data.values, data.leafEncoding);
+    validateArgument(data.format === 'standard-v1', `Unknown format '${data.format}'`);
+    validateArgument(data.leafEncoding !== undefined, 'Expected leaf encoding');
```
These check may appear unecesasry to typescript users, who already get that check from the typing system. But for JS users, this is still needed.

Amxx reviewed (approved):
LGTM

Amxx commented on src/merkletree.ts:
```diff
@@ -0,0 +1,177 @@
+  getMultiProof(leaves: (number | T)[]): MultiProof<HexString, T>;
+  verify(leaf: number | T, proof: HexString[]): boolean;
+  verifyMultiProof(multiproof: MultiProof<BytesLike, number | T>): boolean;
+}
```
We could define dump() here as 
```suggestion
  dump(): MerkleTreeData<T>;
}
```

The benefit would be that it would trigger a typescript error if some some class that extends `MerkleTreeImpl<T>` forgets to include `dump`.

Amxx commented on src/merkletree.ts:
```diff
@@ -0,0 +1,178 @@
+  }
+
+  protected validateValue(valueIndex: number): HexString {
+    invariant(valueIndex >= 0 && valueIndex < this.values.length, 'Index out of bounds');
+    const { value: leaf, treeIndex } = this.values[valueIndex]!;
+    invariant(valueIndex >= 0 && valueIndex < this.values.length, 'Index out of bounds');
```
Duplicated check, probably an error in ada28f1bb405ca688f1a9787a58cf8e5fb870669

Amxx commented on src/merkletree.ts:
```diff
@@ -0,0 +1,178 @@
+    return hashedLeaf;
+  }
+
+  protected leafHash(leaf: number | T): HexString {
```
Fixed

Amxx commented on src/merkletree.ts:
```diff
@@ -0,0 +1,166 @@
+  values: { value: T; treeIndex: number }[];
+};
+
+export interface MerkleTree<T> {
```
Ideally we would have a way to prevent T from being `number`, as this leads to confusion between leaf value and leaf index.

I tried doing:
```typescript
export abstract class MerkleTreeImpl<T, _ = Exclude<T, number>> implements MerkleTree<T> {
```
But it doesn't work as I intended.

ernestognw replied:
Couldn't make it work. I think the best we can do is restrict it at the MerkleTree constructor.

I added an update, it'd be the last thing imo, otherwise LGTM.

Amxx replied:
LGTM

ernestognw commented:
The new "snapshot" files don't convince me but I think it's not a big deal given we don't have too many variations.

I guess we can merge this but I guess is another strong argument in favor of Ava as a testing framework.

frangio commented on src/bytes.ts:
```diff
@@ -1,19 +1,12 @@
-  }
-
-  return a.length - b.length;
+function compare(a: BytesLike, b: BytesLike): number {
```
Does the new function implement the same ordering as the previous one? At first sight I don't think it does. Wondering if this might cause a breaking change.

frangio replied:
I think the previous function was lexicographic ordering and the new one is numerical ordering.

However, it seems we always use this function on byte arrays of length 32, and in this domain the two functions would coincide.

ernestognw replied:
No, it doesn't, but it's related to this:

https://github.com/OpenZeppelin/merkle-tree/issues/30

We may mention it in the `CHANGELOG` but I don't think it's a significant change since people could only use the standardLeaf hash before

ernestognw replied:
> However, it seems we always use this function on byte arrays of length 32, and in this domain the two functions would coincide.

Yes this, "lexicographic"

![image](https://github.com/OpenZeppelin/merkle-tree/assets/33379285/48680019-5af2-4af0-a1b0-37c8faa4827a)

frangio replied:
I don't think it's necessary to mention it in the changelog. My concern was that it would result in tree leaves getting ordered differently. But I don't think this happens. Feel free to resolve if you agree.

Amxx replied:
Its true that both function don't match in the general case, but they match in the case of 32 bytes buffers.

My concern with the previous one is that it doesn't clearly match the solidity version. If we ever get in a situation where both version order differently, my guess is that this new one will be the one that matches solidity's ordering.

frangio commented on src/core.ts:
```diff
@@ -147,18 +136,16 @@ export function isValidMerkleTree(tree: Bytes[]): boolean {
         return false;
       }
-    } else if (!equalsBytes(node, hashPair(tree[l]!, tree[r]!))) {
+    } else if (node !== hashPair(tree[l]!, tree[r]!)) {
```
This is not right. `BytesLike` can be an array type, for which `===` is not the equality that we want. Additionally, for hex strings there are potential issues with inconsistent use of upper and lower case letters. We should keep an `equalsBytes` function, and for strings make sure to lowercase the string before comparing.

Amxx replied:
I updated the code to use compare here. Does that feel ok, of should we be worried of performance issues. I'd rather not have multiple utils doing the same basic thing.

frangio replied:
No concerns about performance. I would prefer `const equalsBytes = (a, b) => compare(a, b) === 0` but up to you.

Amxx replied:
AFAIK, there is only one location where such comparaison is needed. I would avoid adding a definition, and export, and an import for something that is only going to be used once.

ernestognw replied:
I also like `equalsBytes` but I agree with @Amxx I'd rather not reexport it now that it's not used too frequently.

frangio commented on src/core.ts:
```diff
@@ -41,22 +35,22 @@ export function makeMerkleTree(leaves: Bytes[]): Bytes[] {
     index = parentIndex(index);
   }
-  return proof;
+  return proof.map(node => toHex(node));
```
We can remove the `map` and apply the function directly in `push`.

frangio commented on src/core.ts:
```diff
@@ -65,13 +59,11 @@ export interface MultiProof<T, L = T> {
-  if (indices.slice(1).some((i, p) => i === indices[p])) {
-    throw new Error('Cannot prove duplicated index');
-  }
+  validateArgument(!indices.slice(1).some((i, p) => i === indices[p]), 'Cannot prove duplicated index');
```
```suggestion
  validateArgument(indices.slice(1).every((i, p) => i !== indices[p]), 'Cannot prove duplicated index');
```

frangio commented on src/core.ts:
```diff
@@ -97,44 +89,41 @@ export function getMultiProof(tree: Bytes[], indices: number[]): MultiProof<Byte
   return {
-    leaves: indices.map(i => tree[i]!),
-    proof,
+    leaves: indices.map(i => tree[i]!).map(node => toHex(node)),
```
```suggestion
    leaves: indices.map(i => toHex(tree[i]!)),
```

frangio commented on src/simple.ts:
```diff
@@ -0,0 +1,42 @@
+import { MerkleTreeOptions } from './options';
+import { validateArgument } from './utils/errors';
+
+export type SimpleMerkleTreeData = MerkleTreeData<BytesLike> & {
```
I think we will want this to always be `HexString`. For example, arrays of numbers should not be valid in a dump.

```suggestion
export type SimpleMerkleTreeData = MerkleTreeData<HexString> & {
```

Amxx replied:
That works for the `SimpleMerkleTree`, but for the `StandardMerkleTree`, `T extends any[]` may contain crazy things that will end up in dumps.

frangio commented on src/merkletree.ts:
```diff
@@ -0,0 +1,174 @@
+import { MerkleTreeOptions, defaultOptions } from './options';
+import { validateArgument, invariant } from './utils/errors';
+
+export type MerkleTreeData<T extends Exclude<any, number>> = {
```
`Exclude<any, number>` is the same as `any`. I can see that you're doing this to avoid issues with `getProof`. It does not achieve what you want to achieve.

I think you should use `any` and keep the runtime check that you have in the constructor. There are ways to do it statically but they are complex and I don't think it's worth it.

ernestognw replied:
I recall testing it and I'm realizing I got confused because `number` is not compatible with `BytesLike`. Thanks for taking a look.

I see they were removed so solving

frangio commented on src/merkletree.ts:
```diff
@@ -0,0 +1,174 @@
+import { MerkleTreeOptions, defaultOptions } from './options';
+import { validateArgument, invariant } from './utils/errors';
+
+export type MerkleTreeData<T extends Exclude<any, number>> = {
```
Is there a reason why you went with `type` instead of `interface` here?

I don't see any concrete issues so feel free to ignore, but you can use `interface` here, and instead of using `&` in the more specific `*Data` types, you can use `extends`, which allows you to override fields like the `format` one with more narrow types in the same way. I think I prefer interfaces and extending, instead of intersections, whenever that's possible.

frangio reviewed (commented):
Mostly minor comments but also I think one blocker. Feel free to ignore any nitpicks.
